The job **`aarch64-gnu-llvm-20-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/20533450434/job/58988196083) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/58988196083) [(plain)](https://github.com/rust-lang/rust/commit/ed1a5e7970184fb3ae2e7bd5af589f7fb5a2f7b9/checks/58988196083/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

---- [codegen] tests/codegen-llvm/scalable-vectors/debuginfo.rs stdout ----
------rustc stdout------------------------------

------rustc stderr------------------------------
error[E0601]: `main` function not found in crate `debuginfo`
##[error]  --> /checkout/tests/codegen-llvm/scalable-vectors/debuginfo.rs:38:2
   |
38 | }
   |  ^ consider adding a `main` function to `/checkout/tests/codegen-llvm/scalable-vectors/debuginfo.rs`
---
------------------------------------------

error: compilation failed!
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/codegen-llvm/scalable-vectors/debuginfo.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "-O" "-Cdebug-assertions=no" "-Zcodegen-source-order" "--emit" "llvm-ir" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/scalable-vectors/debuginfo/debuginfo.ll" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "--edition=2021" "-C" "debuginfo=2" "-C" "target-feature=+sve"
stdout: none
--- stderr -------------------------------
error[E0601]: `main` function not found in crate `debuginfo`
##[error]  --> /checkout/tests/codegen-llvm/scalable-vectors/debuginfo.rs:38:2
   |
```

</details>

--- a/compiler/rustc_codegen_ssa/src/mir/debuginfo.rs
+++ b/compiler/rustc_codegen_ssa/src/mir/debuginfo.rs
@@ -2,7 +2,7 @@ use std::collections::hash_map::Entry;
 use std::marker::PhantomData;
 use std::ops::Range;
 
-use rustc_abi::{BackendRepr, FieldIdx, FieldsShape, ScalableElt, Size, VariantIdx};
+use rustc_abi::{BackendRepr, FieldIdx, FieldsShape, Size, VariantIdx};
 use rustc_data_structures::fx::FxHashMap;
 use rustc_index::IndexVec;
 use rustc_middle::middle::codegen_fn_attrs::CodegenFnAttrFlags;
@@ -437,16 +437,11 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                 // debugging experience anyway.
                 if operand.layout.ty.is_scalable_vector()
                     && bx.sess().target.arch == rustc_target::spec::Arch::AArch64
-                    && let ty::Adt(adt, args) = &operand.layout.ty.kind()
-                    && let Some(marker_type_field) =
-                        adt.non_enum_variant().fields.get(FieldIdx::from_u32(0))
                 {
-                    let marker_type = marker_type_field.ty(bx.tcx(), args);
+                    let (count, element_ty) =
+                        operand.layout.ty.scalable_vector_element_count_and_type(bx.tcx());
                     // i.e. `<vscale x N x i1>` when `N != 16`
-                    if let ty::Slice(element_ty) = marker_type.kind()
-                        && element_ty.is_bool()
-                        && adt.repr().scalable != Some(ScalableElt::ElementCount(16))
-                    {
+                    if element_ty.is_bool() && count != 16 {
                         return;
                     }
                 }

--- a/compiler/rustc_codegen_ssa/src/mir/debuginfo.rs
+++ b/compiler/rustc_codegen_ssa/src/mir/debuginfo.rs
@@ -431,10 +431,10 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                 // Internally, with an intrinsic operating on a `svint32_t`/`<vscale x 4 x i32>`
                 // (for example), the intrinsic takes the `svbool_t`/`<vscale x 16 x i1>` predicate
                 // and casts it to a `svbool4_t`/`<vscale x 4 x i1>`. Therefore, it's important that
-                // the `<vscale x 4 x i32>` never spills because that'll cause errors during
+                // the `<vscale x 4 x i1>` never spills because that'll cause errors during
                 // instruction selection. Spilling to the stack to create debuginfo for these
-                // intermediate values must be avoided and won't degrade the debugging experience
-                // anyway.
+                // intermediate values must be avoided and doing so won't affect the
+                // debugging experience anyway.
                 if operand.layout.ty.is_scalable_vector()
                     && bx.sess().target.arch == rustc_target::spec::Arch::AArch64
                     && let ty::Adt(adt, args) = &operand.layout.ty.kind()

--- a/tests/ui/scalable-vectors/debuginfo-does-not-spill.rs
+++ b/tests/ui/scalable-vectors/debuginfo-does-not-spill.rs
@@ -1,0 +1,39 @@
+// Compiletest for rust-lang/rust#150419: Do not spill operands to the stack when
+// creating debuginfo for AArch64 SVE predicates `<vscale x N x i1>` where `N != 16`
+//@ edition: 2021
+//@ only-aarch64
+//@ build-pass
+//@ compile-flags: -C debuginfo=2 -C target-feature=+sve
+
+#![crate_type = "lib"]
+#![allow(internal_features)]
+#![feature(rustc_attrs, link_llvm_intrinsics)]
+
+#[rustc_scalable_vector(16)]
+#[allow(non_camel_case_types)]
+#[repr(transparent)]
+pub struct svbool_t(bool);
+
+#[rustc_scalable_vector(4)]
+#[allow(non_camel_case_types)]
+#[repr(transparent)]
+pub struct svbool4_t(bool);
+
+impl std::convert::Into<svbool_t> for svbool4_t {
+    #[inline(always)]
+    fn into(self) -> svbool_t {
+        unsafe extern "C" {
+            #[link_name = "llvm.aarch64.sve.convert.to.svbool.nxv4i1"]
+            fn convert_to_svbool(b: svbool4_t) -> svbool_t;
+        }
+        unsafe { convert_to_svbool(self) }
+    }
+}
+
+pub fn svwhilelt_b32_u64(op1: u64, op2: u64) -> svbool_t {
+    unsafe extern "C" {
+        #[link_name = "llvm.aarch64.sve.whilelo.nxv4i1.u64"]
+        fn _svwhilelt_b32_u64(op1: u64, op2: u64) -> svbool4_t;
+    }
+    unsafe { _svwhilelt_b32_u64(op1, op2) }.into()
+}

